File: api/src/main/java/fr/huiitre/tools/modules/core/admin/infrastructure/PostgresAdminStatsRepository.java
package fr.huiitre.tools.modules.core.admin.infrastructure;

import fr.huiitre.tools.modules.core.admin.application.ports.AdminStatsRepository;
import fr.huiitre.tools.modules.core.admin.application.view.AdminStatsView;
import org.springframework.jdbc.core.JdbcTemplate;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

public class PostgresAdminStatsRepository implements AdminStatsRepository {

    private final JdbcTemplate jdbcTemplate;

    public PostgresAdminStatsRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    @Override
    public AdminStatsView getStats() {
        final String userCountSql = """
                    SELECT
                        COUNT(*) AS total,
                        COUNT(*) FILTER (WHERE is_active = true) AS active
                    FROM tools_core.users
                    WHERE user_type = 'HUMAN'
                """;

        Map<String, Object> counts = jdbcTemplate.queryForMap(userCountSql);
        long totalUsers = ((Number) counts.get("total")).longValue();
        long activeUsers = ((Number) counts.get("active")).longValue();

        final String newUsersSql = """
                    SELECT COUNT(*) FROM tools_core.users
                    WHERE user_type = 'HUMAN'
                    AND created_at >= ?
                """;

        Long newCount = jdbcTemplate.queryForObject(newUsersSql, Long.class, LocalDateTime.now().minusWeeks(1));
        long newUsersThisWeek = newCount != null ? newCount : 0;

        final String perModuleSql = """
                    SELECT m.code, m.name, COUNT(DISTINCT umr.user_id) AS user_count
                    FROM tools_core.module m
                    LEFT JOIN tools_core.user_module_role umr ON m.id = umr.module_id
                    GROUP BY m.id, m.code, m.name
                    ORDER BY user_count DESC
                """;

        List<AdminStatsView.ModuleUserCount> usersPerModule = jdbcTemplate.query(
                perModuleSql,
                (rs, rowNum) -> new AdminStatsView.ModuleUserCount(
                        rs.getString("code"),
                        rs.getString("name"),
                        rs.getLong("user_count")));

        return new AdminStatsView(totalUsers, activeUsers, newUsersThisWeek, usersPerModule);
    }
}

File: api/Modules/Admin/Infrastructure/PostgresAdminStatsRepository.cs
using Dapper;
using Npgsql;
using Tools.ApiCore.Modules.Admin.Application.Dto;
using Tools.ApiCore.Modules.Admin.Application.Ports;

namespace Tools.ApiCore.Modules.Admin.Infrastructure;

// Adaptateur PostgreSQL/Dapper du port IAdminStatsRepository.
public sealed class PostgresAdminStatsRepository(NpgsqlDataSource dataSource) : IAdminStatsRepository
{
    public async Task<AdminStatsDto> GetStatsAsync()
    {
        // Deux résultats en un aller-retour. L'API Java lance trois requêtes séparées ; le
        // décompte hebdomadaire tient dans la première, la répartition par module porte sur
        // d'autres tables et reste distincte.
        const string sql = """
            SELECT COUNT(*) AS TotalUsers,
                   COUNT(*) FILTER (WHERE is_active) AS ActiveUsers,
                   COUNT(*) FILTER (WHERE created_at >= now() - interval '7 days') AS NewUsersThisWeek
            FROM tools_core.users;

            SELECT m.code AS ModuleCode, m.name AS ModuleName,
                   COUNT(DISTINCT umr.user_id) AS UserCount
            FROM tools_core.module m
            LEFT JOIN tools_core.user_module_role umr ON umr.module_id = m.id
            GROUP BY m.id, m.code, m.name
            ORDER BY UserCount DESC, m.name;
            """;

        await using var connection = await dataSource.OpenConnectionAsync();
        await using var results = await connection.QueryMultipleAsync(new CommandDefinition(sql));

        var counts = await results.ReadSingleAsync<UserCountsRow>();
        var perModule = (await results.ReadAsync<ModuleUserCountDto>()).ToList();

        return new AdminStatsDto(
            counts.TotalUsers,
            counts.ActiveUsers,
            counts.NewUsersThisWeek,
            perModule);
    }

    private sealed record UserCountsRow(long TotalUsers, long ActiveUsers, long NewUsersThisWeek);
}
